README.md
## CONTENTS ##
  1. [ABOUT](#ABOUT.md)
  1. [WHAT'S NEW](#WHAT_S_NEW.md)
  1. [LICENSE](#LICENSE.md)
  1. [REQUIREMENTS](#REQUIREMENTS.md)
  1. [CONTACT](#CONTACT.md)
  1. [INSTALLATION](#INSTALLATION.md)
  1. [OPERATION](#OPERATION.md)
    1. [Location Requestor](#Location_Requestor.md)
    1. [Generic NMEA over Bluetooth](#Generic_NMEA_over_Bluetooth.md)
    1. [Nokia Positioning](#Nokia_Positioning.md)
  1. [USAGE](#USAGE.md)
    1. [Logging](#Logging.md)
    1. [Waypoints](#Waypoints.md)
    1. [Track Segments](#Track_Segments.md)
    1. [Display Modes](#Display_Modes.md)
    1. [Destination](#Destination.md)
  1. [SETTINGS](#SETTINGS.md)
    1. [General](#General.md)
    1. [Landmarks](#Landmarks.md)
  1. [TROUBLESHOOTING](#TROUBLESHOOTING.md)
    1. [Known Problems](#Known_Problems.md)
    1. [FAQ](#FAQ.md)
  1. [SCREENSHOTS](#SCREENSHOTS.md)
  1. [ACKNOWLEDGEMENTS](#ACKNOWLEDGEMENTS.md)

# ABOUT #
[pygpslog](http://code.google.com/p/pygpslog/) is a GPS logger for S60
smartphones that uses Nokia's internal location and landmarks API and
thus can co-exist with other navigation applications (e.g. Nokia Maps)
running at the same time.

It's features include
  * Record data in standard [GPX](http://www.topografix.com/GPX) or [OziExplorer](http://www.oziexplorer.com/) format.
  * Easy marking of waypoints (with or without confirmation).
  * Easy marking of six different track segment types (e.g. speed limits).
  * The usual: Data view, Satellite View, Compass View...
  * Simultaneous logging and navigation (e.g. with Nokia Maps).
  * Integration with Nokia's landmarks.

But best see the [Screenshots](README#SCREENSHOTS.md) below.

# WHAT'S NEW #
Please see the [History](History.md) page.

# LICENSE #
> see [LICENSES](LICENSES.md) and [COPYING](COPYING.md)

# REQUIREMENTS #
  * A device running S60 3rd edition.
  * An internal or external GPS device.
  * Python for S60, v1.4.2 or higher. Versions 1.9.x and higher are **not** supported at this time.
> You may get it from [SourceForge](http://sourceforge.net/project/showfiles.php?group_id=154155).
> If you're new to Python on S60 it's probably better to read the
> [PyS60 Wiki](http://wiki.opensource.nokia.com/projects/PyS60) first.
  * The [LocationRequestor](http://www.iyouit.eu/portal/Software.aspx) module (**see below!**).
  * The [Landmarks](http://www.iyouit.eu/portal/Software.aspx) module (**see below!**).

  * All installation files have to be [Symbian Signed](https://www.symbiansigned.com) by yourself in order to be really useful. Please see the [DOCUMENTATION](README#INSTALLATION.md)

The LocationRequestor module above is packaged with an UID that cannot be
used for Symbian's [Open Signed](https://www.symbiansigned.com/app/page/public/openSignedOnline.do)
process. Therefore I'm making a
[re-packaged version](http://code.google.com/p/pygpslog/downloads/list?can=2&q=label%3ARecommended+LocationRequestor)
available here.

Unfortunately, the Landmarks module, as of Mar. 7, 2008, contains a
few bugs that prevent it from working with pygpslog. Until I receive word
from the authors, you may download a
[patched version](http://code.google.com/p/pygpslog/downloads/list?can=2&q=label%3ARecommended+Landmarks)
from this site.

In fact, the LocationRequestor and Landmarks modules are not _absolutely_
required to run pygpslog, but the functionality will be severely limited.
Please see [below](README#OPERATION.md).

# CONTACT #
Please leave your comments on the [Contact](Contact.md) wiki page. I will be notified by
e-mail automagically ;-)

# INSTALLATION #
  1. Install Python for S60 from [SourceForge](http://sourceforge.net/project/showfiles.php?group_id=154155).
  1. If any old version of the LocationRequestor or Landmarks modules are installed, please uninstall them
  1. Get the [LocationRequestor](http://code.google.com/p/pygpslog/downloads/list?can=2&q=label%3ARecommended+LocationRequestor) module.
  1. Get the [Landmarks](http://code.google.com/p/pygpslog/downloads/list?can=2&q=label%3ARecommended+Landmarks) module.
  1. Get the [latest version](http://code.google.com/p/pygpslog/downloads/list?q=label:Featured) of pygpslog.
  1. If you're not yet, make yourself familiar with [Symbian Signed](https://www.symbiansigned.com). The easiest way by now seems to be [Open Signed Online](https://www.symbiansigned.com/app/page/public/openSignedOnline.do).
  1. Sign LocationRequestor, Landmarks and PyGpsLog. The capabilities required are `ReadUserData, WriteUserData, NetworkServices, LocalServices, UserEnvironment, ReadDeviceData, WriteDeviceData, Location` and `SwEvent`. See SymbianSigned.
  1. Install to your device
  1. Done!

Almost: I'd strongly recommend you to download, sign, and install those additional modules:
  * [e32jext](http://code.google.com/p/pygpslog/downloads/list?can=2&q=label%3ARecommended+e32jext) from this site.
  * [envy](http://prdownloads.sourceforge.net/pyed/envy_3rd_1_0_4_unsigned.sis?download). Read [this post](http://discussion.forum.nokia.com/forum/showthread.php?t=122897&highlight=envy) for a description.
  * [uikludges](http://uikludges.googlecode.com/svn/trunk/uikludges.SIS) from http://code.google.com/p/uikludges/.
  * [miso](http://pdis.hiit.fi/pdis/download/miso/miso-1.92-s60_3rd_dev.sis) from http://pdis.hiit.fi/pdis/download/miso/ or [misty](http://cyke64.googlepages.com/misty150_unsigned_dev_certfree.SIS) from http://cyke64.googlepages.com/.
> These modules give you additional features like a CPU graph, more meaningful dialog menus, vibrating alarm and prevent pygpslog from being accidentally closed by pressing the red key.

I know the signing process is tedious, but this application requires `Location`
capabilities (`SwEvent` might be needed later). To make things worse,
Symbian Signed seems to be in the process of changing their signing procedures.
As of now (Mar. 7, 2008) you're not able to get your own developer certificate
without a valid Publisher ID. But please check yourself at
[Symbian Signed](https://www.symbiansigned.com).

In the meantime I'll try to keep some SymbianSigned instructions [updated](SymbianSigned.md).
Please feel free to comment and contribute.

If you _absolutely_ want to avoid Symbian Signed _and_ you have and _external
Bluetooth_ GPS device, you may skip the installation of LocationRequestor and
Landmarks. In this case _you will not be able_ to share the GPS device
with other applications, and, you cannot use your phone's landmarks in any
way.

# OPERATION #
pygpslog can operate in three different modes. Each mode has it's advantages
and disadvantages. The default mode depends on what you installed during
the [installation process](#INSTALLATION.md).

### Location Requestor ###
(Preferred method, uses 3rd Party Python module)
#### Pro's: ####
  * GPS can be shared with other applications.
  * Very extensive data available.
  * Works with external or internal GPS devices.
  * Continues to work, even if Bluetooth connection is lost.
#### Con's: ####
  * Requires the [LocationRequestor](http://www.iyouit.eu/portal/Software.aspx) module.
  * Must be Symbian signed

### Generic NMEA over Bluetooth ###
#### Pro's: ####
  * Most extensive (and probably precise) data available.
  * Can be used without Symbian Signed.
  * No additional modules required.
#### Con's ####
  * Requires an external Bluetooth GPS.
  * GPS cannot be shared with Nokia Maps, Navigator, mgMaps or others.

### Nokia Positioning ###
(uses standard Python module)
#### Pro's: ####
  * GPS can be shared with other applications.
  * Works with external or internal GPS devices.
  * No additional modules required.
#### Con's: ####
  * Severely limited data (no satellite info, ...).
  * Must be Symbian signed.

# USAGE #

### Logging ###
The `'Yes'` (Green) key starts and stops the GPS logger. Every restart
creates a new log file in the [configured](#SETTINGS.md) data directory (default:
`C:\Data\GpsLog` or `E:\Data\GpsLog`).

The `'C'` (Backspace) key pauses the logging operation.

Logging can be started automatically on program startup (see [SETTINGS](#SETTINGS.md)).

PyGpsLog can store it's log files in two different formats:
##### GPX #####
[GPX](http://www.topografix.com/GPX) is a format used by a
[lot of devices and applications](http://www.topografix.com/gpx_resources.asp), including
[GoogleEarth](http://earth.google.com), and can be easily manipulated.
Additionally, PyGpsLog uses it's [own extensions](Screenshots#Extended_Format.md)
to store additional data for later analysis.

##### OziExplorer #####
[OziExplorer](http://www.oziexplorer.com/) is a powerful GPS mapping software. The format
used for their `.plt` track files is very compact and by
[far enough](Screenshots#OziExplorer.md) to store basic trackpoint data.

So, if you're low on memory use this format.

##### GPSBabel #####
One word for [GPSBabel](http://www.gpsbabel.org/): Depending on the information you need,
it doesn't really matter which format you use. You can alway use
[GPSBabel](http://www.gpsbabel.org/) to convert from one format into almost any other.

From [version 1.3.5 beta](http://www.gpsbabel.org/downloads.html) on, GPSBabel can
create Nokia LMX landmark files that can be sent via Bluetooth to your device. So I'm
using it to create my Landmarks in GoogleEarth, convert them to LMX and send them
to my E65.

### Waypoints ###
The `'#'` key inserts a waypoint into the log file.
For both, [GPX](#GPX.md) and [OziExplorer](#OziExplorer.md) formats, a separate log file
is created that holds the marked waypoints. The files are named
`'<Logfile>-wpt.gpx'` and `'<Logfile>.wpt'` respectively.

For [GPX](#GPX.md) files [there's an option](#General.md) to merge the track- and
waypoint-logfiles into a single file. This is done on request only, because
according to [GPX](#GPX.md)-syntax, waypoints have to appear before tracks, and
re-writing a large log file takes _considerable_ time on device
(see [this issue](http://code.google.com/p/pygpslog/issues/detail?id=4)).

If the Landmarks module is [installed](#INSTALLATION.md), and the option is
[activated](#Landmarks.md), a Nokia landmark with the current time stamp
is added.

### Track Segments ###
There are six different segment-marker categories. Their default names are:
`"Speed 30", "Speed 50", "Speed 60", "Speed 70", "Speed 80"` and `"Speed 100"`.

If you want to configure these markers yourself, please read about UserIcons.

The segments are activated by pressing one of the number keys from `'4'` to `'9'`.
To end a segment press the same key again or press the '`*'`-key.

If you press a different key from `'4'` to `'9'` while a segment is active,
the current segment will be closed and a new segment opened.

If the Landmarks module is [installed](#INSTALLATION.md), and the option is
[activated](#Landmarks.md), a Nokia landmark is added, that marks the
center between the start and the end of the segment, has a coverage
radius of half the distance and intersects the end points.

By default these landmarks are directional, i.e. notifications only take
place if you're going roughly in the same direction as the landmark.
If you want to change this, you have to save your tracks in GPX format,
[extract and edit](extractmarks.md) the segments and then write the landmarks
back to the device.

### Display Modes ###
To save battery, information display is usually only active if logging is switched on.

You can switch on GPS information in idle mode by pressing the `'Select'` key (or
`'Enter'` key, joystick button).

The left and right arrow key or the `'1'` and `'3'` keys switch back and forth
between the different modes.
  1. Overview - Log File Name and General GPS
  1. Compass and Speed
  1. Landmarks (if [installed](#INSTALLATION.md))
  1. Satellite View
  1. Power saving mode (log file writing only)

The `'2'` key switches the application in full screen mode. This especially
useful in night mode.

Night and day mode can be toggled with the `'0'` key. Night mode will
be automatically active the the application is started after 7pm or before
5am (should be 9am, but never mind ;-) )

The `'C'` (Backspace) key also freezes the GPS display.

### Destination ###
PyGpsLog provides a simple way of navigating, a little bit like Nokia's
own Navigator application.

The `Destination` menu allows you to select a destination, either from your
landmarks (if [installed](#INSTALLATION.md)) or as decimal coordinates.

Once a destination is selected it's distance and estimated arrival time is shown
on the Overview and Compass pages. In the landmarks list it will be shown on top.

On the Compass and Satellite pages there will be two dot helping you navigate.
A **blue** marker shows the absolute heading and a **red** marker will show heading
relative to your current heading (or view direction, when going forward).
Here's a [screnshot](Screenshots#Compass.md).

# SETTINGS #
## General ##
  * **Keep Backlight On**: Keeps the devices' backlight on while logging or GPS display is active. Default: **off**.

History.md
# Version History #
### Legend ###
+ Added feature

- Bugfix

~ Changed functionality

## 0.3.1 - Mar. 30, 2009 ##
  * - Don't use undocumented positioning features for PyS60 versions >= 1.4.5 (fixes [Issue 8](https://code.google.com/p/pygpslog/issues/detail?id=8)).
  * ~ Add `'E:\resource`' to sys.path in case modules are installed on `'E:`'.

## 0.3.0 - Mar. 23, 2008 (Happy Easter!) ##
  * + Write marked waypoints to a separate GPX or OziExplorer file.
  * + Option to merge finished GPX tracks and waypoints.
  * + User configurable track segment markers.
  * + Optionally insert manually entered destinations as waypoints/landmarks.
  * + Added simple keyboard help.
  * + [extractmarks](extractmarks.md) can now create bi-directional landmarks.
  * - Fixed a bug in arrival time display.
  * - Switch off vibration while charging. May hang some devices (like the E65).
  * - Bugfixes in internal settings.
  * ~ Added a progress dialog to the `'Set Icons/Categories'` option.
  * ~ Don't set landmark icons for global categories.
  * ~ Relaxed parsing of destination coordinates.

## 0.2.3 - Mar. 21, 2008 ##
  * + Select categories to be displayed (see [settings](README#Landmarks.md)).
  * + Show running average speed on compass page, if a destination is selected.
  * - Landmark access optimized. Still eats a lot of CPU, so keep you charger ready.
  * ~ Switched to the GPLed [Tulliana 2](README#ACKNOWLEDGEMENTS.md) icon set

## 0.2.2 - Mar. 18, 2008 ##
  * + Added support for [directional track segments](extractmarks.md)
  * + Added support for [audible and vibrating alarms](README#Landmarks.md).
  * - Fixed a bug that caused wrong categories for track segments

## ~~0.2.1 - Mar. 17, 2008~~ ##
**Revoked due to a bad track segment marking bug**
  * + Support for UserIcons in landmarks
  * - Fixed slow application loading.
  * - Don't show average speed if logfile is closed
  * - Switch to normal screen on landmark edit (may crash Python).
  * ~ Lowered the icon quality to reduce installed size.

## 0.2.0 - Mar. 14, 2008 ##
  * + Simple Navigation (choose a destination)
  * + Display GPS data without starting a log file.
  * + Night mode
  * + Full screen mode with battery meter
  * + CPU usage meter
  * + Use `envy`-module to prevent accidental red-key-exit.
  * + Ask for restart on GPS failure
  * - Waypoint marking fixed
  * - Fix crash on lost GPS
  * ~ Optimized landmark drawing

## 0.1.7 - Mar. 07, 2008 ##
Initial public version.

src/gpsloglm.py
try:    import landmarks
except: landmarks = None
import  gpslogutil

INF = 1e300**2 # These might not work in future python versions
NAN = INF/INF

if landmarks != None:
  import os, e32, appuifw
  import lmsearch

  LANDMARK_DB = u"file://e:gpslog.ldb"
  LANDMARK_DB = None

  ICONDIR  = r"C:\Data\Others" # (not e32.in_emulator() and str(os.path.split(appuifw.app.full_name())[0])) or "C:\Python"
  ICONFILE = unicode(os.path.join(ICONDIR,"gpsloglm_%s.mif" % appuifw.app.uid()))
  #MKICON   = lambda ico: appuifw.Icon(ICONFILE, ico, ico)
  MKICON   = lambda ico: appuifw.Icon(*ico)
  
  def findIcons(): # really: Install icons ;-)
    mydir = "." # str(os.path.split(appuifw.app.full_name())[0])
    srcicon = os.path.join(mydir, "gpsloglm.mif")
    if not os.path.exists(srcicon):
      srcicon = os.path.join(os.path.dirname(__file__), "gpsloglm.mif")
    # if e32.in_emulator(): mydir = r"C:\Python"
    if not os.path.exists(ICONFILE) or os.stat(srcicon).st_mtime > os.stat(ICONFILE).st_mtime:
      import shutil
      shutil.copy2(srcicon, ICONFILE)

  findIcons()
  
  ICON_DEF = [ 
    ("Default",  0x4000), ("GPS Log",  0x4000), ("Speed",     0x400A),
    ("Speed 30", 0x4002), ("Speed 50", 0x4004), ("Speed 60",  0x4006),
    ("Speed 70", 0x4008), ("Speed 80", 0x400A), ("Speed 100", 0x400C),
  ]
  ICONS    = dict([ (n, (ICONFILE, id, id)) for n, id in ICON_DEF ])

  WPT_LM_ATTR = 0x003F & ~(landmarks.EDescription)
                # landmarks.EAllAttributes &~... makes API panic (and app terminate)!
  
  waypointCache = {}

  def addUserIcons(icons):
    global ICON_DEF, ICONS
    for name in icons:
      ico = icons[name].mifIcon()
      if ico != None:
        ICON_DEF += [ (name, None) ]
        ICONS[name] = ico

  def OpenDb(uri=LANDMARK_DB, create=False):
    if uri != None:
      raise NotImplementedError, "It doesn't make sense anyway. Use Categories!"
    else:
      return landmarks.OpenDefaultDatabase()

  def ClearCache():
    while waypointCache:
      del waypointCache[waypointCache.keys()[0]]
  
  def ReadLm(db, id):
    # no caching here: we crash on closing the cached lm's
    return db.ReadLandmark(id)
    
  def ReadPartialLm(db, id, attr=WPT_LM_ATTR, fields=[]):
    oattr, ofields = db.PartialReadParameters()
    db.SetPartialReadParameters(attr, fields)
    lm = db.ReadPartialLandmark(id)
    db.SetPartialReadParameters(oattr, ofields)
    return lm
    
  def ReadLmWpt(db, id):
    if id in waypointCache:
      return waypointCache[id]

    # lm = ReadPartialLm(db, id)
    lm = ReadLm(db, id)
    pos = lm.GetPosition()
    rad = lm.GetCoverageRadius()
    attr = { "categories": lm.GetCategories() }
    if rad != None: attr["radius"] = rad
    hdg = None
    if lm.IsPositionFieldAvailable(landmarks.EPositionFieldHeading):
      hdg = float(lm.GetPositionField(landmarks.EPositionFieldHeading))
    # HACK ALERT:  lmx format doesn't allow for import or export of heading
    elif lm.IsPositionFieldAvailable(landmarks.EPositionFieldBuildingTelephone):
      hdg = lm.GetPositionField(landmarks.EPositionFieldBuildingTelephone)
      if hdg.startswith("hdg:"): hdg = float(hdg[4:])
    wpt = (gpslogutil.Waypoint([ lm.GetLandmarkName(), pos[0], pos[1], pos[2],
                                 None, None, hdg, None, None, None, None,
                                 pos[3], pos[4], attr ] ), lm.GetIcon())
    lm.Close()
    waypointCache[id] = wpt
    return wpt

      
  def CreateCat(name, icon=None):
    c   = landmarks.CreateLandmarkCategory()
    c.SetCategoryName(unicode(name))
    if icon != None and type(icon) != tuple:
      icon = ICONS[icon]
    if icon != None:
      c.SetIcon(*icon)
    db  = OpenDb()
    cm  = db.CreateCategoryManager()
    id  = cm.AddCategory(c)
    c.Close(); cm.Close(); db.Close()
      
  def SearchLm(crit, max=-1, asWpt=False, db=None):
      
    if db != None: sdb = db
    else:          sdb = OpenDb()
    
    try:
      ids = lmsearch.Landmarks(crit).findIds(sdb)
      if not asWpt:
        res = ids
      else:
        res = [ ReadLmWpt(sdb, id) for id in ids ]
    finally:
      if db == None:
        sdb.Close()

    return res

  def SetCatIcons(name): # This version gives us a KErrLocked
    db   = OpenDb()
    cm   = db.CreateCategoryManager()
    cid  = cm.GetCategory(unicode(name))
    if cid != landmarks.KPosLmNullItemId:
      cat = cm.ReadCategory(cid)
      if cat.GlobalCategory() != landmarks.KPosLmNullItemId:
        return # ignore Global categories
      icon = cat.GetIcon()
      cat.Close()
      if icon != None:
        ids  = SearchLm(lmsearch.Category(cid), db=db)
        done = [] # dupes possible, then db is locked (though lm is closed :-( )
        for id in ids:
          if id in done: continue
          RETRY = 10
          for i in range(RETRY):
            try:
              lm = ReadLm(db, id)
              lm.SetIcon(*icon); db.UpdateLandmark(lm); lm.Close()
              done += [id]
              break
            except SymbianError, exc:
              if exc[0] != -22 or i >= RETRY-1: raise # KErrLocked
              e32.ao_sleep(0.2)
    cm.Close()
    db.Close()

  def CreateLm(wpt, name=None, desc=None, cat=[], radius=None, edit=False, fields=[], fakehdg=False):
    if not type(cat) in [list, tuple]:
      cat = [ cat ]
    lm = landmarks.CreateLandmark()
    if name == None:
      name = wpt[0]
    lm.SetLandmarkName(unicode(name))
    if desc != None:
      lm.SetLandmarkDescription(unicode(desc))
    if radius != None:
      lm.SetCoverageRadius(radius)
    for c in cat:
      if c != landmarks.KPosLmNullGlobalCategory: # :-))) You have to love it!
        lm.AddCategory(c)

    if wpt.hdg != None:
      fields += [ (landmarks.EPositionFieldHeading, str(wpt.hdg)) ]
      # HACK ALERT:  lmx format doesn't allow for import or export of heading
      if fakehdg:
        fields += [ (landmarks.EPositionFieldBuildingTelephone, "hdg:"+str(wpt.hdg)) ]

    for f in fields:
     lm.SetPositionField(f[0], unicode(f[1]))

    db = OpenDb()
  
    if cat:
      cm = db.CreateCategoryManager()
      if cat[0] != landmarks.KPosLmNullGlobalCategory: # :-)))))))
        c  = cm.ReadCategory(cat[0])
        icon = c.GetIcon()
        if icon != None:
          lm.SetIcon(*icon)
        c.Close()
      cm.Close()
    # wpt.alt = None
    alt, hacc, vacc = [ (v != None and v) or NAN for v in (wpt.alt, wpt.hacc, wpt.vacc)]
    lm.SetPosition(wpt.lat, wpt.lon, alt, hacc, vacc)
    id = db.AddLandmark(lm)
    lm.Close()
    if id and edit:
      screen = appuifw.app.screen
      try:
        appuifw.app.screen = "normal"
        id = db.ShowEditDialog(id, landmarks.ELmkAll, landmarks.ELmkEditor, False)
      finally:
        appuifw.app.screen = screen
      
    db.Close()
  
  def NearestLmClassic(lat, lon, max=16, maxdist=-1.0, cat=-1, db=None):
    if lat == None or lon == None:
      return []

    if cat != -1 and cat != None:
      if type(cat) not in [list, tuple]: cat = [cat]
    else: cat = []

    nrst = lmsearch.Nearest((lat, lon), maxdist, True) 

    if len(cat) != 0:
      sc = [ (nrst, lmsearch.Category(c)) for c in cat ]
    else:
      sc = [ nrst ]

    res  = SearchLm(sc, max=max, asWpt=True, db=db)

    while sc: del sc[0]

    return res

  def NearestLm(lat, lon, max=16, maxdist=-1.0, cat=-1, db=None):
    if lat == None or lon == None:
      return []

    if cat != -1 and cat != None:
      if type(cat) not in [list, tuple]: cat = [cat]
    else: cat = []

    nrst = lmsearch.Nearest((lat, lon), maxdist, True) 

    if len(cat) == 0:
      res = SearchLm(nrst, max=max, asWpt=True, db=db)

    else:
      # We want the result ordered by distance
      if db != None: sdb = db
      else:          sdb = OpenDb()


      near = lmsearch.Landmarks(nrst).findIds(db=sdb)

      if near:
        usecompost = lmsearch.Landmarks.USE_COMPOSITE
        lmsearch.Landmarks.USE_COMPOSITE = True
        sc  = [ (lmsearch.ItemIds(near), lmsearch.Category(c)) for c in cat ]
        ids = lmsearch.Landmarks(sc).findIds(db=sdb)
        ids = [ id for id in near if id in ids ]
        if max > 0: ids = ids[:max]
        res = [ ReadLmWpt(sdb, id) for id in ids ]
        del near, ids
        while sc: del sc[0]
        lmsearch.Landmarks.USE_COMPOSITE = usecompost
      else:
        res = []

      if db == None:
        sdb.Close()


    del nrst

    return res
  

  class LandmarkSettings(gpslogutil.GpsLogBaseSettings):

    catnames = property(lambda self: [ c[0] for c in self.categories ])
    catids   = property(lambda self: [ c[1] for c in self.categories ])

    ##########################################################################
    def __init__(self, _desc_list=None, _fname=None):
    
      self.__loadCat()

      wptnames = [ unicode(c[0]) for c in self.categories ]
      desc = [
        (("uselm",   "Use Landmarks", True), "number",   [], 1),
        (("wptlm",   "Store Waypoints as Landmarks"), "combo",   [u"on", u"off"], u"on"),
        (("lmedit",  "Edit New Landmarks"), "combo",   [u"on", u"off"], u"on"),
        (("wptcat",  "Category for Waypoints"), "combo",  wptnames , u"(None)"),
        (("marklm",  "Store Markers as Landmarks"), "combo",   [u"on", u"off"], u"on"),
        (("radius",  "Landmark Search Radius (km)"), "number",   [], 10),
        (("warncat", "Notify near Category"), "combo",  wptnames , u"(None)"),
        (("warnrad", "Distance for Notfication (m)"), "number",  [] , 10),
        (("warnvib", "Vibrate on Notification"), "combo",   [u"on", u"off"], u"on"),
        (("upd",     "Update interval (s)"), "number",   [], 3),
        (("smico",   "Small Icons"), "combo",   [u"on", u"off"], u"off"),
        (("lmico",   "Show Nokia Icons"), "combo",   [u"on", u"off"], u"off"),
       #(("dispcat", "Display Categories", True), "text",   [], u"[]"),
        (("dispcat", "Display Categories", True), "text",   [], u"None"),
      ]
      self.ONOFFS   = [ "wptlm", "marklm", "lmedit", "lmico", "smico", "warnvib" ]
    
      gpslogutil.GpsLogBaseSettings.__init__(self, desc, "gpsloglm.settings")
      self.__apply()
      
    ##########################################################################
    def __apply(self):
      for attr in self.ONOFFS:
       setattr(self, attr, getattr(self, attr)  == "on")

      if self.radius == 0: self.radius = -1
      self.radius = float(self.radius)
      
      if not self.wptcat in self.catnames:
        self.wptcat = self.categories[0][0]
      self.wptcat = self.cat(self.wptcat)[1]

      if not self.warncat in self.catnames:
        self.warncat = self.categories[0][0]
      self.warncat = self.cat(self.warncat)[1]

      self.dispcat = eval(self.dispcat)
      # kill removed categories
      if self.dispcat != None:
        self.dispcat = [ c[1] for c in self.categories if c[1] in self.dispcat]

    ##########################################################################
    def __cnvt(self):
      for attr in self.ONOFFS:
        setattr(self, attr,  (getattr(self, attr)  and "on") or "off")

      self.radius = int(self.radius)
      if self.radius < 0: self.radius = 0

      try:
        self.wptcat = self.cat(self.wptcat)[0]
      except:
        self.wptcat = self.categories[0]
    
      try:
        self.warncat = self.cat(self.warncat)[0]
      except:
        self.warncat = self.categories[0]
      
      self.dispcat = unicode(repr(self.dispcat))

    ##########################################################################
    def __loadCat(self):
      self.categories = [ ("(None)", landmarks.KPosLmNullGlobalCategory), # :-)) == 0
                          ("(Create New)", -1) ]
      db  = OpenDb()
      cm  = db.CreateCategoryManager()
      cit = cm.CategoryIterator(landmarks.ECategorySortOrderNameAscending) # :-)))
      id  = cit.Next()
      while id != landmarks.KPosLmNullItemId: # :-))))
        cat = cm.ReadCategory(id)
        self.categories += [ (cat.GetCategoryName(), id) ]
        cat.Close()
        id = cit.Next()
      cit.Close()
      cm.Close()
      db.Close()
    
    ##########################################################################
    def cat(self, idx):
      if isinstance(idx, (str, unicode)): # name as index
        return [ c for c in self.categories if unicode(c[0]) == unicode(idx)][0]
      else: # id as index
        return [ c for c in self.categories if c[1] == idx][0]

    ##########################################################################
    def lmeditor(self):
      screen = appuifw.app.screen
      try:
        appuifw.app.screen = "normal"
        rc, id, db = landmarks.ShowSelectLandmarkDialog(landmarks.KPosLmNullItemId)
        if rc == 0: return
        db.ShowEditDialog(id, landmarks.ELmkAll, landmarks.ELmkEditor, True)
        db.Close()
      finally:
        appuifw.app.screen = screen
      
    ##########################################################################
    def showcat(self):
      # we're in execute_dialog i.e. we're in unicode world now!
      # and, 'None' means all, '[]' means None ;-)
      dispcat = eval(self.dispcat)
      allcat    = self.categories[2:]
      if dispcat != None:
        choicecat = [ c for c in allcat if not c[1] in dispcat ]
        choicecat.sort()
        if len(choicecat) == 0: # something happened...
          dispcat = None;  self.dispcat = "None"

      if dispcat == None:
        appuifw.note(u"All categories already displayed", "error")
        return
      
      choices = [ unicode(c[0]) for c in choicecat ]
      sel = appuifw.multi_selection_list(choices, style="checkbox", search_field=True)
      if not sel: return # Cancel
      
      dispcat += [ choicecat[c][1] for c in sel ]
      if len(dispcat) == len(allcat): dispcat = None
      self.dispcat = str(repr(dispcat))

    ##########################################################################
    def hidecat(self):
      # we're in execute_dialog i.e. we're in unicode world now!
      dispcat = eval(self.dispcat)
      if dispcat == []:
        appuifw.note(u"No categories displayed", "error")
        return
      if dispcat == None:
        dispcat = [ c[1] for c in self.categories[2:]]

      choicecat = [ c for c in self.categories if c[1] in dispcat]
      choicecat.sort()

      choices = [ unicode(c[0]) for c in choicecat ]
      sel = appuifw.multi_selection_list(choices, style="checkbox", search_field=True)
      if not sel: return # Cancel

      hidden = [ choicecat[c][1] for c in sel ]
      rest = [ c for c in dispcat if not c in hidden ]
      rest.sort()
      self.dispcat = str(repr(rest))

    ##########################################################################
    def addDefCat(self):
      prog = gpslogutil.ProgressDialog()
      prog.show()
      tot = len(ICON_DEF)-1; i = 1
      try:
        for nm, icon in ICON_DEF[1:]:
          prog.text = nm; prog.progress = i * 100 / tot; i += 1
          if not nm in self.catnames:
            CreateCat(nm, nm)
          if nm != "Speed": # sorry for the special case
            SetCatIcons(nm)
      except Exception, exc:
        prog.hide()
        appuifw.note(u"An error occurred: %s" % str(exc), "error")
        
      del prog

      appuifw.note(u"Please close and re-open Landmark settings", "conf")
    
    ##########################################################################
    def execute_dialog(self):
      self.uselm = False
      try:
        self.__init__() # Categories may have been changed by other app!
        self.__cnvt()

        menu = [
          (u"Landmark Editor", self.lmeditor),
          (u"Add Display Categories", self.showcat),
          (u"Remove Disp. Categories", self.hidecat),
          (u"Set Icons/Categories", self.addDefCat),
        ]
        #if e32.in_emulator() or __file__.lower() == r"e:\python\gpslog\gpsloglm.py":
        #  menu += [
        #  ]

        oldcat  = self.wptcat
        oldwarn = self.warncat

        try:
          gpslogutil.GpsLogBaseSettings.execute_dialog(self, menu)
        except:
          self.reset()
          self.__init__()
          self.save()
          appuifw.note(u"Settings have been reset (new version?)", "error")
          gpslogutil.GpsLogBaseSettings.execute_dialog(self, menu)
        
        if self.lmico == "on":
          rc = appuifw.query(u"Icons are an experimental feature! Enable?", "query")
          if not rc: self.lmico = "off"
            
        self.__apply()
      
        if self.wptcat == -1:
          if not self.newCategory():
            self.wptcat = oldcat
        elif self.warncat == -1:
          if not self.newCategory(wptcat=False):
            self.warncat = oldwarn
        
        self.save()
      finally:
        self.uselm = True

    ##########################################################################
    def newCategory(self, wptcat=True):
      while True:
        ncat = appuifw.query(u"New Category", "text", u"GPS Log")
        if ncat == None or ncat not in self.catnames:
          break
        appuifw.note(u"Category exists", "error")
      
      if ncat == None:
        return False

      body = appuifw.app.body
      exit = appuifw.app.exit_key_handler
      menu = appuifw.app.menu

      lock = e32.Ao_lock()
      self.tmp_sel = None
      try:
        en = []
        for nm, id in ICON_DEF:
          en += [ (unicode(nm), MKICON(ICONS[nm])) ]
        lb  = None
        def selected():
          self.tmp_sel = lb.current()
          lock.signal()
        lb = appuifw.Listbox(en, selected)
        if ncat in ICONS:
          idx = [ i[0] for i  in ICON_DEF ].index(ncat)
          lb.set_list(en, idx)
        appuifw.app.exit_key_handler = lock.signal
        appuifw.app.menu = [ (u"Ok", selected), (u"Cancel", lock.signal) ]
        appuifw.app.body = lb
        lock.wait()
      finally:
        appuifw.app.body = body
        appuifw.app.menu = menu
        appuifw.app.exit_key_handler = exit

      icon = None
      if self.tmp_sel != None: icon = ICON_DEF[self.tmp_sel]

      CreateCat(ncat, icon)
      
      if wptcat: self.wptcat  = id
      else:      self.warncat = id
      self.categories += [ (ncat, id) ]
      self.save()
      self.__init__()

      return True

    ##########################################################################
    def save(self):
      self.__cnvt()
      gpslogutil.GpsLogBaseSettings.save(self)
      self.__apply()


else: # landmarks == None:
  class NoLandmarkSettings(object):
    def __init__(self):
      self.uselm = False
    def save(self):
      pass

  LandmarkSettings = NoLandmarkSettings


def close():
  if landmarks:
    ClearCache()
    landmarks.ReleaseLandmarkResources()
